src/main/java/com/github/princesslana/slothbot/middleware/PermissionMiddleware.java

```diff
@@ -0,0 +1,31 @@
+package com.github.princesslana.slothbot.middleware;
+
+import com.github.princesslana.slothbot.Config;
+import com.github.princesslana.slothbot.Discord;
+import com.google.gson.JsonObject;
+import disparse.discord.smalld.Event;
+import disparse.discord.smalld.permissions.Permission;
+import disparse.discord.smalld.permissions.PermissionUtils;
+import java.util.function.BiFunction;
+
+public class PermissionMiddleware implements BiFunction<Event, String, Boolean> {
+  @Override
+  public Boolean apply(Event event, String commandName) {
+    if (Config.getPublicCommands().contains(commandName)) {
+      return true;
+    }
+    if (PermissionUtils.computeAllPerms(event).contains(Permission.ADMINISTRATOR)) {
+      return true;
+    }
+    JsonObject json = event.getJson().getAsJsonObject("d");
+
+    var possibleGuildId = Discord.getGuild(json);
+    var roles = Discord.getRoles(json);
+    return possibleGuildId
+        .map(
+            guildId ->
+                roles.stream()
+                    .anyMatch(role -> Config.getModerator().containsModerator(guildId, role)))
```

Comment:
I think it's a bit cleaner to pass Moderator into a constructor here, limiting the wiring up of everything to App.main.